ports/esp32/machine_timer.c

```diff
@@ -251,31 +349,84 @@ static mp_obj_t machine_timer_init_helper(machine_timer_obj_t *self, mp_uint_t n
         mp_raise_ValueError(MP_ERROR_TEXT("hard Timers are not implemented"));
     }
 
+    // Calculate the period setting
+    uint64_t vperiod = 0;
+    uint64_t hperiod = 0;
     #if MICROPY_PY_BUILTINS_FLOAT
     if (args[ARG_freq].u_obj != mp_const_none) {
-        self->period = (uint64_t)(machine_timer_freq_hz() / mp_obj_get_float(args[ARG_freq].u_obj));
+        hperiod = (uint64_t)(machine_timer_freq_hz() / mp_obj_get_float(args[ARG_freq].u_obj));
+        // 1s as 1,0000,000 microseconds gives us the virtual timer tick for freq=n, 1M/n = m microseconds
+        vperiod = 1000000.0 / mp_obj_get_float(args[ARG_freq].u_obj);
```

Comment:
I kind of get why you've chosen to calculate both together here, but it's a bit awkward to read back. How about something like this?

```c
uint32_t clock_freq;
const uint32_t VIRT_FREQ = 1000000; // 1s in microseconds

#if MICROPY_PY_BUILTINS_FLOAT
if (args[ARG_freq].u_obj != mp_const_none) {
    clock_period = self->is_period ? VIRT_FREQ : machine_timer_freq_hz();
    self->period = clock_period / mp_obj_get_float(args[ARG_freq].u_obj);
}
#else
   // repeat pattern for integer version
#endif
[...]

if (self->period == 0) {
    mp_raise_msg(&mp_type_ValueError, MP_ERROR_TEXT("period too small for clock frequency %u", clock_freq));
}
``` 

?